### ISSUE TITLE
This app will not work with future versions of macOS [macOS Warning]

### ISSUE THREAD

**Issue by andrewcrook** [iina/iina#3563]
**System and IINA version:**

- macOS 12.0.1 (21A559)
- IINA: 1.2.0 Build 129

***** Issue ******
First time using IINA and with a YouTube URL
I got this warning 
Doesn’t seem to be logs

<img width="747" alt="Screenshot 2021-12-07 at 15 23 17" src="https://user-images.githubusercontent.com/803618/145061658-6cd0f8ef-97b3-4c71-b2cc-c10d8d608af9.png">

**Comment by byGits** [comment]
IINA was the best video player for many years (even better than VLC)

Unfortunately it seems to be abandoned now by original devs...

**Comment by lhc70000** [comment]
It's not abandoned. Hopefully I can release the plugin system earlier next year so people can update youtube-dl (yt-dlp) without rebuilding/releasing IINA. The plugin system works now but it lacks docs and testing.

**Comment by CarterLi** [comment]
> It's not abandoned. Hopefully I can release the plugin system earlier next year so people can update youtube-dl (yt-dlp) without rebuilding/releasing IINA. The plugin system works now but it lacks docs and testing.

Are youtube-dl things that important? Even more important than those crashing, compatibility issues? How many people watch youtube videos using IINA?

**Comment by byGits** [comment]
Not at all, I dont even use youtube-dl and I never will.

When I said abandoned it was because iina has been crashing for all tall videos for almost a year now, and last update was on Februrary...

I get other random crashes too sometimes, which forced me to go back to VLC.

Hope I can get back to iina soon, I love iina when it doesn't crash!

-------- Original Message --------

**Comment by byGits** [comment]
Glad I was wrong!
I dont use youtubedl, said abandoned because its been constantly crashing with no updates for almost a year.

Thank you for not bandoning iina, hope it will get stable again soon.

-------- Original Message --------

**Comment by byGits** [comment]
> It's not abandoned. Hopefully I can release the plugin system earlier next year so people can update youtube-dl (yt-dlp) without rebuilding/releasing IINA. The plugin system works now but it lacks docs and testing.

Glad I was wrong!

I don't use yt-dl, said abandoned because its been constantly crashing with no updates for almost a year.

Thank you for not abandoning iina, hope it will get stable again soon.

(Edit: Sorry for re-posting, my previous reply had issues)

**Comment by lhc70000** [comment]
@CarterLi

Many people watching youtube videos using IINA doesn't mean that I'm obligated to serve the people (for free), although I'm happy to try my best. I'm not an independent developer or a paid software engineer. I'm a stressed Ph.D. student, and updating IINA won't help my course and project deadlines and make my life easier.

It's unfortunate if a developer stops maintaining a project. But since it's open-source, you still have a lot of things to do. You can try to contact any of the organization members to request repo access if you want to contribute. If failed, you are free to fork the project and merge the updates back probably in the future. The build script is available for the public. But when you do this, please be thankful to all contributors, regardless of previous or current, rather than blaming them. They wrote these code when they didn't have to and made this project possible.

If you know there's any PR fixing crashing and compatibility issues, could you give me a link? I'm almost finished with this semester and may have some time to merge them.

**Comment by lhc70000** [comment]
@byGits 

If you happen to know any issue mentioning the crash, could you give me a link? Sorry that I didn't review the issues for almost a year and it would be great to find some critical ones.

**Comment by byGits** [comment]
> @byGits
> 
> If you happen to know any issue mentioning the crash, could you give me a link? Sorry that I didn't review the issues for almost a year and it would be great to find some critical ones.

Sure and thanks again for not abandoning IINA:

This one still happens to me constantly (most tall/vertical/portrait videos cause a crash):
https://github.com/iina/iina/issues/3364

I'm not exactly sure why some other random freezes/crashes happen yet, but I'll try to investigate and report.

**Comment by CarterLi** [comment]
> and updating IINA won't help my course and project deadlines and make my life easier.

Yeah. However you contributors need to file an announcement at least saying that you are busy currently or something. Your keeping silence does make other [contributors](https://github.com/iina/iina/pulls) and other IINA lovers sad. Submitting PR doesn't make their life easier too.

> You can try to contact any of the organization members to request repo access if you want to contribute

Really? Can I request this repo access?

**Comment by lhc70000** [comment]
@CarterLi 

IINA needs motivated and experienced long-term contributors, and I really want meet someone who want to do some maintaining work (such as reply to the issues and review the PRs in addition to writing new code).

If you use Telegram, could you contact `uiryuu` (@alejx on GitHub) or `lhc70000` (me, but I'm not sure if I have enough time)? We can discuss more and I believe @alejx can answer basic questions about this project.

If you have any existing PR, please give me a link. If you have experience with JavaScript front-end projects, please tell me because the plugin system still needs help.

**Comment by lhc70000** [comment]
@byGits

Thanks a lot, but seems that I can't reproduce it even with the provided sample files like in #3400. Could you provide a working sample?

Also, I'm on Monterey. Does this still happen on Monterey or is Big Sur-only?

**Comment by CarterLi** [comment]
> @CarterLi
> 
> IINA needs motivated and experienced long-term contributors, and I really want meet someone who want to do some maintaining work (such as reply to the issues and review the PRs in addition to writing new code).
> 
> If you use Telegram, could you contact `uiryuu` (@alejx on GitHub) or `lhc70000` (me, but I'm not sure if I have enough time)? We can discuss more and I believe @alejx can answer basic questions about this project.
> 
> If you have any existing PR, please give me a link.

[I used to have one](https://github.com/iina/iina/pull/3539), but I had lost motivation maintaining it because I didn't It would be merged.

Instead I'm maintaining my fork. You can test the HDR feature with other fixes made by other contributors

https://github.com/iina-plus/iina/releases/tag/v1.2.0-13

> If you have experience with JavaScript front-end projects, please tell me because the plugin system still needs help.

Yes I'm a front-end developer. But to be honest I don't have motivation working with youtube things.

**Comment by lhc70000** [comment]
Do you use Telegram or you prefer other messaging apps? I can find @low-batt on Telegram and I think it's better to have a better place to discuss more.

> Yes I'm a front-end developer. But to be honest I don't have motivation working with youtube things.

That's great. I was talking about the whole JavaScriptCore-based ecosystem. Please feel free to explore the `plugin-system` branch if interested.

![image](https://user-images.githubusercontent.com/8478049/145150740-a30ea908-1766-4bc2-a9c1-2cceff8d5032.png)

**Comment by low-batt** [comment]
YES!! @lhc70000 has returned to us from his extended vacation on a beautiful tropical island with no internet connection (that stressed Ph.D. student cover story is a good one). Let me start with this...

THANK YOU to @lhc70000, the other core developers, the extended contributors and the IINA users that have submitted issues. You did such a good job on IINA that you got this committed Internet lurker to create a GitHub account and start proposing fixes / enhancements to IINA.

For people that don't like to read long posts:
I am available now to help the core developers put out an IINA 1.2.1 release.
I am ready to rebase / adapt my current proposed changes as needed.
I am ready to help out with other tasks.

At this point a bug squashing release is overdue.

My story is this... I encountered a few glitches in IINA. Plowed though the open IINA issues and found some of those had already been reported a couple times. Heard mumblings about the core developers being overloaded. So I decided to try and help out the core developers and propose fixes for those issues. I found the code so good that in most cases only a minor tweak was required to fix the problem. In reading through the issues I noticed some other problems that seemed like something I could track down and fix. Here it is months later and I've not managed to get to the end of my list of issues to try and address. Darn users keep reporting new issues.

All of my proposed changes are in PRs submitted to the IINA repository. In many of the associated issues I have detailed my analysis of the issue and how I fixed it, hoping to make it easy for reviewers to spot any errors. I have just started helping @CarterLi with his fork. 

I've been focusing on these kind of issues:
- Crashes
- CPU Loops
- Memory Leaks
- Adherence to Apple Human Interface Guidelines

As well as a few issues that could be considered minor enhancements and some of the simple but tedious work like updating the copyrights, eliminating warnings, etc. I've been avoiding issues that I felt required input from the core developers.

I am a very experienced software engineer with expertise in many different technologies... none of which are being used by IINA. I've never worked on UIs, never mind the macOS Cocoa frameworks. So my submissions require scrutiny. Critical brutal mean reviews are welcome.

@lhc70000 Thoughts on getting started? A GitHub 1.2.1 milestone?

I have a "IINA 1.2.1" markdown document in the style of IINA release notes where I was keeping track of my own proposed changes. There are a number of other contributors that have proposed useful enhancements. I've not had a chance to look closely into them. 

I can tell you a lot about the situation with defects that really need to be addressed. I have posted proposed fixes for many of them. I have also been running full builds of the dependencies using `compile.rb`. I have yet to get a run where all of the dependencies build successfully. I can do another run and report on the current situation.

IINA is supposed to be built with Apple's latest stuff, but CarterLi encountered a compilation failure with my changes and `#available` under Xcode 13.1. We are still unsure what that is all about, but I found the Xcode 13.1 release notes document a regression with `#available`. It did not match what we saw, but it could be the problem is worse than what was documented in the release notes. A fix is apparently coming in 13.2.

I'll checkin on Telegram.

**Comment by low-batt** [comment]
On the issue reported by @andrewcrook...

I believe this from the [macOS Monterey 12.0.1 Release Notes](https://developer.apple.com/documentation/macos-release-notes/macos-12_0_1-release-notes) explains alert:

> **Python**
> **Deprecations**
>
>   - If an app uses Python 2.7, macOS now triggers an alert indicating that the developer must update the app to ensure it will work in future versions of macOS. (80221011)

youtube-dl uses Python 2.
yt-dlp requires Python 3.

Apple starting warning they intend to stop shipping scripting languages with macOS a while back. This is from the [macOS Catalina 10.15 Release Notes](https://developer.apple.com/documentation/macos-release-notes/macos-catalina-10_15-release-notes):

>**Scripting Language Runtimes**
>**Deprecations**
>
>  - Scripting language runtimes such as Python, Ruby, and Perl are included in macOS for compatibility with legacy software. Future versions of macOS won’t include scripting language runtimes by default, and might require you to install additional packages. If your software depends on scripting languages, it’s recommended that you bundle the runtime within the app. (49764202)
>
> - Use of Python 2.7 isn’t recommended as this version is included in macOS for compatibility with legacy software. Future versions of macOS won’t include Python 2.7. Instead, it’s recommended that you run python3 from within Terminal. (51097165)

If IINA continues to embed youtube-dl or yt-dlp then it may have to embed Python if Apple pulls it from macOS.

**Comment by andrewcrook** [comment]
> > **Scripting Language Runtimes**
> > **Deprecations**
> > 
> > * Scripting language runtimes such as Python, Ruby, and Perl are included in macOS for compatibility with legacy software. Future versions of macOS won’t include scripting language runtimes by default, and might require you to install additional packages. If your software depends on scripting languages, it’s recommended that you bundle the runtime within the app. (49764202)


Thanks

I did read this a couple of years ago this was happening I forgot because I expected them to do it in Big Sur.

**Comment by lhc70000** [comment]
For youtube-dl/yt-dlp, my plan is to rewrite the whole online media part as [a plugin](https://github.com/iina/iina-plugin-ytdl), so this will not be a problem soon. Embedding a youtube-dl binary in the main app is too annoying - we must release a new version if these dependencies changed, including building a production version, testing, prepare the release notes, etc.

**Comment by lhc70000** [comment]
@low-batt I sent you an invitation. @saagarjha said that he will review your commits in the weekend. Thanks again for your contribution! 

@CarterLi please tell me your preferred way of contact when you have time. I think the HDR support is an important feature and we'd like to discuss what you want to contribute in the future.

**Comment by saagarjha** [comment]
Yeah, I’m a bit busy with work right now but I’ll try my best to take a look at the crash fixes this weekend. @low-batt if you have PRs that you think should get priority, feel free to list them here and I’ll look at them first.

@CarterLi: I’m not happy with how we’ve been running IINA, but I do want to mention that we’ve certainly not been silent on the issue: #2686 was written a while back. What’s new is that we’ve gotten new interest in someone actually taking on the maintenance burden, and while I’ll be the first to admit we’ve been unreasonably slow to respond to it this is something we’d like to fix.

As @lhc70000 mentioned above, HDR support is something we’d be interested in adding to IINA and we’d love to have you comfortable enough to contribute your changes upstream. At the moment, though, we’ll probably focus on getting the backlog of crashes and core feature support under control first. We understand that a “plugin system” is not really a feature people have been asking for but it is critical in allowing us to offload maintenance work.

**Comment by CarterLi** [comment]
> please tell me your preferred way of contact when you have time.

Well, I'm living in mainland China. Because the various networking issue, my main way of contact is wechat and I don't think any contributor of this repo are using it.

I'd prefer using Github `Discussions` feature. If no, I'd rather using PR comments.

> I think the HDR support is an important feature and we'd like to discuss what you want to contribute in the future.

I'm trying to port my HDR code. Should I port it to the plugin-system branch or develop branch?

**Comment by CarterLi** [comment]
> I’m not happy with how we’ve been running IINA, but I do want to mention that we’ve certainly not been silent on the issue: #2686 was written a while back.

I must say that I missed it. I suggest you should pin it.

**Comment by low-batt** [comment]
@saagarjha Picky reviews welcome. Don't hold back.

I contend that this issue should be marked as a high priority:
Excessive CPU consumption when opening the Playlist Panel #3162

Because in some cases it is not just consuming CPU, it is looping contacting the streaming server.
As continuously contacting a server could get your IP banned I think this one is a priority.

I have posted in the issue my analysis of the problem and details on my proposed fix which is in PR: #3477

I believe this same problem is responsible for issues #3041, #3194, #3341 and #3411.

I will look through my PRs tomorrow and create a list ordered by priority.

**Comment by low-batt** [comment]
With respects to updating library dependencies...

I have submitted PR iina/homebrew-mpv-iina#15 to the homebrew-mpv-iina repository containing a commit to update mpv to [v0.34.1](https://github.com/mpv-player/mpv/releases/tag/v0.34.1). This update contains fixes for OpenGL issues as well as others that look significant.

The closed issue #3327 shows a build problem with `libgnutls` triggering this failure under macOS 10.11:
```text
dyld: Symbol not found: _clock_gettime
  Referenced from: IINA.app/Contents/MacOS/../Frameworks/libgnutls.30.dylib
  Expected in: /usr/lib/libSystem.B.dylib
```
Not sure what to do about that one. See issue for details.

Need to make sure Xcode does not report any warnings with updated libraries such as the one shown in #3496.

Some of the issues updating depedencies should fix:
mpv 0.34.1 #3472, #3463, #3460
libgmp 6.2.1_1 #3503
libass 0.15.2 #3222

Desirable to get the PR #3592, the fix for the "Code Signature Invalid" crash merged as upgrading mpv makes this crash more likely to occur.

**Comment by low-batt** [comment]
Closing. IINA 1.3.0 has been released and it includes [yt-dlp](https://github.com/yt-dlp/yt-dlp) with python embedded as now required by macOS.
IINA 1.3.0 can be downloaded from the official [IINA](https://iina.io/) website.

The warning had to do with Apple's plans to remove python from macOS. That has now occurred. From the [macOS Monterey 12.3 Release Notes](https://developer.apple.com/documentation/macos-release-notes/macos-12_3-release-notes):
> ### Python
> **Deprecations**
>
> Python 2.7 was removed from macOS in this update. Developers should use Python 3 or an alternative language instead. (39795874)

Thus the use of the yt-dlp bundle that includes python.
